forte/data/extractor/feature.py

```diff
@@ -0,0 +1,108 @@
+            list of features.
+        pad_id (int or List):
+            a single integer or a nested list of integer representing <PAD>.
+        dim (int):
+            Total number of dimensions for the data. `dim` is always >= 1.
+            If the data is a list of value, the `dim` should be 1 and this
+            feature is called base feature.
+        """
+        self.pad_id: Union[int, List] = pad_id
+        self.dim: int = dim
+        self.base_feature = dim == 1
+        self.data = None  # Only base feature has actual data
+        self.mask = None
+
+        self._parse_sub_features(data)
+        self._validate_input()
+
+    def _validate_input(self):
+        assert (self.is_base_feature() and self.data is not None or
+                (not self.is_base_feature()
+                 and self.data is None
+                 and self.sub_features is not None))
+        assert type(self.pad_id) == int or type(self.pad_id) == list
+        assert self.dim >= 1
+
+    def _parse_sub_features(self, data):
+        if self.is_base_feature():
+            self.data = data
+        else:
+            self.sub_features: List = []
+            for sub_data in data:
+                self.sub_features.append(
+                    Feature(sub_data, self.pad_id, self.dim - 1))
+
+        self.mask = [1] * len(data)
+
+    def is_base_feature(self) -> bool:
+        return self.base_feature
+
+    def get_sub_features(self) -> List['Feature']:
+        assert not self.is_base_feature(), \
+            "Base feature does not have sub features"
+        assert self.dim > 1, \
+            "Non-base feature should have as least 2 dimension"
+
+        return self.sub_features
+
+    def get_len(self) -> int:
+        return len(self.data) if self.is_base_feature() else \
+            len(self.sub_features)
+
+    def pad(self, max_len: int):
+        assert self.get_len() <= max_len, \
+            "Feature length should not exceed given max_len"
+
+        for i in range(max_len - self.get_len()):
+            if self.is_base_feature():
+                self.data.append(self.pad_id)
+            else:
+                self.sub_features.append(Feature([], self.pad_id, self.dim - 1))
+            self.mask.append(0)
+
+    def unroll(self, need_pad: bool = True) -> Tuple[List[Any], List[Any]]:
+        if not need_pad:
+            return self.data, []
+
+        if self.is_base_feature():
+            return self.data, [self.mask]
+        else:
+            unroll_features: List = []
+            sub_stack_masks: List = []
+
+            for feature in self.get_sub_features():
+                sub_unroll_features, sub_masks = feature.unroll()
+
+                for i in range(self.dim - 1):
+                    if i == len(sub_stack_masks):
+                        sub_stack_masks.append([])
+                    sub_stack_masks[i].append(sub_masks[i])
+
+                unroll_features.append(sub_unroll_features)
+
```

Comment:
If `unroll` is the main call to get the content of this feature, does it mean that at each call we will repeat all computation?

Reply:
Yes. But the caller only need to call this once and I think if we will call unroll multiple time, probably it should be the caller's responsibility to cache the result from unroll?

Reply:
Why don't you build all the features after initialization, and in the future simply call the value?

Reply:
We cannot do that, because one `Feature` represents a single data point (a single instance in the dataset). 

1) During the initialization, it does not know how to do the padding because padding needs to know the `max_len` but this information can only be given by the caller because only caller has the view of a batch of data. Caller will check all the length for each data point in the batch and pass `max_len` to the `Feature::pad` so that each `Feature` can do the padding and after that, the `unroll` can help get the actual value.

2) We cannot redesign and use `Feature` represents a batch of data points because we have the semantic that `extractor` will output `Feature` and doing that means that `extractor` will also need to extract a batch of data at a time and this will prevent us from havingsingle-data-point as cache granularity (as we have to cache a batch in this case)